Update CommandPyImp.cpp comment setParameters()

J-Dunn opened the pull request:
setParameters() is NOT a "setter" method. There is not setter method and this attribute does not behave as a python dictionary. 

Since the whole point is to provide python-like behaviour, this is a serious bug which only remains unexplored due to the very limited use context of this attribute.

Until this is fixed a clear note is added to avoid confusion.

<!-- Include a brief summary of the changes. -->

<!--
The FreeCAD community thanks you for your contribution!
By creating a Pull Request you agree to the contributing policy. The complete policy can be found in the root of the source tree (CONTRIBUTING.md) or at https://github.com/FreeCAD/FreeCAD/blob/main/CONTRIBUTING.md

This template provides guidance on creating a PR that can be reviewed and approved as quickly as possible. Comments may be safely deleted.

Unless you know exactly what you're doing, please leave the checkbox 'Allow edits by maintainers' enabled.  This will allow maintainers to help you.
-->

## Issues
<!-- link to individual issues this PR closes by referencing the issue number (e.g., fixes #1234, closes #4321). -->

## Before and After Images
<!-- If your proposed changes affect the FreeCAD GUI, add before and after screenshots -->



<!--  Notes on the PR Review Process

The following section describes what the maintainers consider when reviewing your Pull Request.  These items may not require you to take any action.  This information is provided for context. Understanding what we consider will help you prepare your request for speedy approval.

You can find additional documentation about these guidelines in the [Developers handbook](https://freecad.github.io/DevelopersHandbook).

Alignment (Does the PR align with the goals and interests of the project?)
  - Does the PR have at least one issue linked, which this PR closes?
  - Has the conversation on the PR and related issue(s) reached consensus?
  - If the PR affects the GUI, is the Design Working Group (DWG) aware and have they had time to review and comment?
  - If the PR affects the GUI, did the contributor include before/after images?
  - If the PR affects standards and workflow, is the CAD Working Group (CWG) aware and have they had time to review/comment?

Impact (Does the change affect other parts of the project?)
  - Has the impact on documentation been considered and appropriate action taken?
  - Has the impact on translation been considered appropriate action taken?
  - Will the PR affect existing user documents?

Code Quality (Is code well-written and maintainable?)
  - Does the PR warrant a review by the Code Quality Working Group (CQWG)?
  - Does the change include tests?
  - Is the PR rebased on the current main branch with unnecessary commits squashed?

Release (Are there considerations related to release timing?)
  - Has the PR been considered for backporting to the latest release branch?
  - Have the release notes been considered/updated?
  -->

J-Dunn commented:
Example of unexpected results using Parameters in python: 

```
        paramDict = {"X": 0, "Y": 0}
        HxCommandList.append(Path.Command(arc_cmd, paramDict))
        HxCommandList[-1].Parameters = {}
```
Instead of assigning an empty dictionary, the list line does nothing:  it ADDS an empty dict. 

https://forum.freecad.org/viewtopic.php?p=823333

phaseloop commented:
I would suggest to change this behaviour for what is expected from a setter - and eventually patch particular operations if they rely on old behaviour.

J-Dunn commented:
Yes , I think this should work correctly. 

It would be nice to have input from @sliptonic  who knows the  usage of this better from the practical python end, but it seems that this is exclusively used in single constructor calls where gcode commands are added to the list object.

I don't think it was ever intended NOT to be a setter, and as far as I have seen it is never intentionally used as an addParameter() as opposed to a setter.

3x380V commented:
@J-Dunn, please do not open PR like this one. Create an issue instead, if you are unable to fix code yourself, then close this.

sliptonic commented:
I agree with @3x380V about the PR/Issue.   Also ping @yorikvanhavre on the issue.  He was more directly involved in original implementation of the Path Command C++ code and will likely have more insight than I.

J-Dunn commented:
> I agree with @3x380V about the PR/Issue. Also ping @yorikvanhavre on the issue. He was more directly involved in original implementation of the Path Command C++ code and will likely have more insight than I.

I know you were not really involved in the cpp code, I said "@sliptonic who knows the usage of this better from the practical python end."   I thought you may be able to comment on how it is use , not how well it was implemented. 

Are you aware of anywhere where this is used outside of commandlist.append(Path.Command(....)) type usage ?

J-Dunn commented:
> Create an issue instead, if you are unable to fix code yourself, then close this.

I'd be fairly confident I can fix it, it's question of whether there is a will to see it fixed. 
If no one says "it was supposed to function that way : don't touch" I'll add another commit. 
If there's objections, this comment at least needs to be in there.

3x380V commented:
No, it does not. It just adds more noise and solves nothing. I guess you have enough attention already, so things will certainly sort out.

J-Dunn commented:
To apply exptected behaviour for a setter, it seems enough to clear Parameters before adding with the existing code. 

```
void CommandPy::setParameters(Py::Dict arg)
{
    PyObject* dict_copy = PyDict_Copy(arg.ptr());
    PyObject *key, *value;
    Py_ssize_t pos = 0;
    getCommandPtr()->Parameters.clear();
    while (PyDict_Next(dict_copy, &pos, &key, &value)) {
        std::string ckey;
....
```

This does not show any problems on testing with existing paths I have, though that is only a superficial test. 

This all depends on how this is currently being used in python in Path WB.  Since no one seems to have even noticed in the years this has been extensively used and tested , that tends to confirm my initial impression that Parameters is only ever set via the Path.Command() constructor and never actually used as a setter to change Parameters to a new dict object. 

It would be nice to have confirmation from one of the cadres (maybe @wmeyer ) of the PyCXX code that this is indeed an oversight and/or from someone from the Path WB end who has more extensive knowledge of how this is actually used in the python code. 

Since these getters and setters are often only called implicitly , it's hard to be sure to check all current usage.

3x380V commented:
History of that file itself shows that `CommandPy::setParameters` implementation is just a verbatim copy from `CommandPy::PyInit` constructor as you can see from the initial commit https://github.com/FreeCAD/FreeCAD/commit/3c9d77aa3c1afd5774b271f3c57b841d0d5473a1 merged a decade ago. Your impression is therefore almost certainly correct :)
Btw, you probably wanted to ping @wwmayer, but that is unlikely to work atm... Fortunately @yorikvanhavre is principal author here and he is responding quite regularly (with understandable delay of course).

J-Dunn commented:
Thanks for digging that out.  It looks like this has been quietly hiding there for a decade. 

  I've updated the PR with my alleged bug fix and removed the now redundent comment.

J-Dunn commented:
While we are about python-like behaviour, should print(commandlist[-1]) produce something which looks like a python dict?
https://forum.freecad.org/viewtopic.php?t=96554#p823740

Current output:
`print(commandlines[i])
Command G0 [ F:0 X:0 Y:0 Z:11 ]`

IMO the Parameters part of that output should be a curly bracketed, comma separate list of key:value pairs. Currently it is neither list nor dict and no commas. 

The following, however, prints correctly since it references a python dict. 
 `commandlines[i].Parameters`

I can add this to the current PR : 
```
using namespace Path;

// returns a string which represents the object e.g. when printed in python
std::string CommandPy::representation() const
{
    std::stringstream str;
    str.precision(5);
    str << "Command ( ";
    str << getCommandPtr()->Name;
    str << ", {";
    for (std::map<std::string, double>::iterator i = getCommandPtr()->Parameters.begin();
         i != getCommandPtr()->Parameters.end(); ++i) {
        std::string k = i->first;
        double v = i->second;
        str << "\'" << k << "\':" << v;
        if ( i != std::prev(getCommandPtr()->Parameters.end())  ) str << ", ";
    }
    str << "} )";
    return str.str();
}

```

>>> params = {"X":0.0 , "Y":2.1}
>>> cmd = Path.Command("G2",params)
>>> params
{'X': 0.0, 'Y': 2.1}
>>> cmd
Command ( G2, {'X':0, 'Y':2.1} )

J-Dunn commented:
There is a remaining problem with Command, Command.Parameters :

```
>>> cmd = Path.Command("G0",{"Z":0.0})
>>> cmd
Command ( G0, {'Z':0} )
>>> cmd.Parameters["x"] = 1
>>> cmd.Parameters["y"] = 2
>>> cmd.Parameters["f"] = 3.0
>>> cmd
Command ( G0, {'Z':0} )
```
Since this accesses Parameters as dict Command.setParameters is not called. 
However, 
```
>>> cmd.Parameters
{'Z': 0.0, 'x': 1, 'y': 2, 'f': 3.0}

```
Now let's call the setter 
```
>>> cmd.Parameters = cmd.Parameters
>>> cmd
Command ( G0, {'F':3, 'X':1, 'Y':2, 'Z':0} )
```
Is this a consequence of the parameters_copy_dict bugfix for the mega memleak a few years back?

I do not think direct addition to Command.Parameters is done anywhere due to simple, limited use of Path.Command in FC but it caught me out when trying to add a new gcode param after the initial call to the constructor FreeCAD (exlcusively does.)

J-Dunn commented:
The Pixies seem to be having a union work to rule operation. Maybe we should pay them more. 
Automatic rebuilds have been stuck to 24h.

J-Dunn commented:
https://github.com/FreeCAD/FreeCAD/issues/21038

J-Dunn commented:
Cleaning up this mess. 

New PR uniquely for correcting setParameters to actually act as a setter.
Update CommandPyImp.cpp setParameters
#21039

Associated "issue" : PATH CommandPy::setParameters isn't a setter. #21038
